src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetcher.java

```diff
@@ -144,50 +180,135 @@ public GHRepository repositoryFor(GitHubProject project, GitHub github) throws I
-   * @param github of type {@link GitHub} connection to GitHub API.
-   * @return {@link GHCommit} with the project information.
-   * @throws IOException occurred during REST call to GitHub API.
+   * Clones a repository of a specified project.
+   *
+   * @param project The project.
+   * @return A local repository.
+   * @throws IOException If something went wrong while cloning the repository.
    */
-  public Optional<GHCommit> firstCommitFor(GitHubProject project, GitHub github)
-      throws IOException {
-    List<GHCommit> commits = commitsFor(project, github);
-    if (commits.isEmpty()) {
-      return Optional.empty();
+  public LocalRepository localRepositoryFor(GitHubProject project) throws IOException {
+    Objects.requireNonNull(project, "On no! Project is null!");
+
+    LocalRepository repository = localRepositories.get(project);
+    if (repository == null) {
+      repository = loadLocalRepositoryFor(project);
+      localRepositories.put(project, repository);
     }
-    return Optional.ofNullable(commits.get(commits.size() - 1));
+
+    return repository;
   }
 
   /**
-   * Get a list of commits which comes after a specific date.
-   * 
-   * @param date which is used as boundary.
-   * @param project of type {@link GitHubProject}, which holds the project information.
-   * @param github of type {@link GitHub} connection to GitHub API.
-   * @return sub-list of type {@link GHCommit}.
-   * @throws IOException occurred during REST call to GitHub API.
+   * Loads a repository of a specified project.
+   *
+   * @param project The project.
+   * @return A local repository.
+   * @throws IOException If something went wrong.
    */
-  public List<GHCommit> commitsAfter(Date date, GitHubProject project, GitHub github)
-      throws IOException {
-    List<GHCommit> subList = new ArrayList<>();
+  private LocalRepository loadLocalRepositoryFor(GitHubProject project) throws IOException {
+    Objects.requireNonNull(project, "On no! Project is null!");
 
-    for (GHCommit commit : commitsFor(project, github)) {
-      if (commit.getCommitDate().before(date)) {
-        break;
-      }
-      subList.add(commit);
+    LocalRepositoryInfo info = localRepositoriesInfo.get(project.url());
+    if (info == null) {
+      Path repositoryPath = base
+          .resolve(project.organization().name())
+          .resolve(project.name());
+
+      info = new LocalRepositoryInfo(
+          repositoryPath, Date.from(Instant.now()), project.url());
+    }
+
+    if (Files.isRegularFile(info.path())) {
+      LOGGER.warn("{} is a file but it should be a directory, let's remove it", info.path());
+      Files.delete(info.path());
+    }
+
+    Optional<Repository> repository = openRepository(info.path());
+    if (!repository.isPresent()) {
+      Files.deleteIfExists(info.path());
+      clone(project, info.path());
+      repository = openRepository(info.path());
     }
 
-    return subList;
+    if (!repository.isPresent()) {
```

Comment:
Is there a way to combine this and the above if condition. I understand the significance of checking it twice. But could something else be written?

Reply:
Those two check, which look similar, in fact check a bit different things. The first check checks if an an existing repository can be loaded. If not, they we do a little cleanup and, clone a fresh repository, and then try to open it again. So, the second check checks if the fresh repository can be opened.

Reply:
I meant can we move the lines #L228 and #L229 inside the `openRepository()`. So that we do not have to check if `!repository.isPresent()` twice.

In the end if !repository.isPresent() is still true then throw the exception.

Does it make sense?

Reply:
The lines #L228 and #L229 should not be moved to the `openRepository()`. That would mean that we always want to clone a fresh repository. But we want to re-use existing ones. The second `!repository.isPresent()` could be moved inside the first `if` but I wanted to avoid nested ifs.

Reply:
This is a fallback logic. We first check if we have a valid repository. If not, we try to fetch a fresh one (fallback). If we still can't open the repository, then we break.

Reply:
ok